src/random_ray/iteration.cpp

```diff
@@ -0,0 +1,660 @@
+    if (volume == 0.0) {
+      continue;
+    }
+
+    int material = random_ray::material[sr];
+
+    double sr_fission_source_old = 0;
+    double sr_fission_source_new = 0;
+
+    for (int e = 0; e < negroups; e++) {
+      int64_t idx = (sr * negroups) + e;
+      double nu_sigma_f = data::mg.macro_xs_[material].get_xs(
+        MgxsType::NU_FISSION, e, nullptr, nullptr, nullptr, t, a);
+      sr_fission_source_old += nu_sigma_f * random_ray::scalar_flux_old[idx];
+      sr_fission_source_new += nu_sigma_f * random_ray::scalar_flux_new[idx];
+    }
+
+    fission_rate_old += sr_fission_source_old * volume;
+    fission_rate_new += sr_fission_source_new * volume;
+  }
+
+  double k_eff_new = k_eff_old * (fission_rate_new / fission_rate_old);
+
+  return k_eff_new;
+}
+
+// Apply a few sanity checks to catch obvious cases of numerical instability.
+// Instability typically only occurs if ray density is extremely low.
+void instability_check(int64_t n_hits, double k_eff, double& avg_miss_rate)
+{
+  double percent_missed = ((random_ray::n_source_regions - n_hits) /
+                            static_cast<double>(random_ray::n_source_regions)) *
+                          100.0;
+  avg_miss_rate += percent_missed;
+
+  if (percent_missed > 10.0) {
+    warning(fmt::format(
+      "Very high FSR miss rate detected ({:.3f}%). Instability may occur. "
+      "Increase ray density by adding more rays and/or active distance.",
+      percent_missed));
+  } else if (percent_missed > 0.01) {
+    warning(fmt::format("Elevated FSR miss rate detected ({:.3f}%). Increasing "
+                        "ray density by adding more rays and/or active "
+                        "distance may improve simulation efficiency.",
+      percent_missed));
+  }
+
+  if (k_eff > 10.0 || k_eff < 0.01 || !(std::isfinite(k_eff))) {
+    fatal_error("Instability detected");
+  }
+}
+
+// Enforces restrictions on inputs in random ray mode.  While there are
+// many features that don't make sense in random ray mode, and are therefore
+// unsupported, we limit our testing/enforcement operations only to inputs
+// that may cause erroneous/misleading output or crashes from the solver.
+void validate_random_ray_inputs()
+{
+  // Validate tallies
+  ///////////////////////////////////////////////////////////////////
+  for (auto& tally : model::tallies) {
+
+    // Validate score types
+    for (auto score_bin : tally->scores_) {
+      switch (score_bin) {
+      case SCORE_FLUX:
+      case SCORE_TOTAL:
+      case SCORE_FISSION:
+      case SCORE_NU_FISSION:
+      case SCORE_EVENTS:
+        break;
+      default:
+        fatal_error(
+          "Invalid score specified. Only flux, total, fission, nu-fission, and "
+          "event scores are supported in random ray mode.");
+      }
+    }
+
+    // Validate estimator types
+    if (tally->estimator_ != TallyEstimator::ANALOG) {
+      fatal_error("Invalid estimator specified. Only analog estimators are "
```

Comment:
Why analog? Aren't these all track length tallies? Analog only makes sense to me for actual collisions, but maybe I'm misinterpreting something.

Reply:
Great point! There was an issue that was preventing this, but managed to track it down. Now tracklength/collision/analog are all supported, although they all do the same thing (basically akin to tracklength tallying). I've updated the code and added a note in the docs that says they all are basically going to look something like MC tracklength tallies and that there's no difference.